keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/CheckScheduler.kt

```diff
@@ -43,25 +47,49 @@ class ResourceCheckScheduler(
   @Scheduled(fixedDelayString = "\${keel.resource-check.frequency:PT1S}")
   fun checkResources() {
     if (enabled) {
-      log.debug("Starting scheduled validation…")
-      publisher.publishEvent(ScheduledCheckStarting)
+      log.debug("Starting scheduled resource validation…")
+      publisher.publishEvent(ScheduledResourceCheckStarting)
 
       val job = launch {
         resourceRepository
-          .nextResourcesDueForCheck(Duration.ofMinutes(resourceCheckMinAgeMinutes), resourceCheckBatchSize)
-          .forEach { (_, name, apiVersion, kind) ->
-            launch {
-              resourceActuator.checkResource(name, apiVersion, kind)
-            }
+          .launchForEachItem {
+            resourceActuator.checkResource(it.name, it.apiVersion, it.kind)
           }
       }
 
       runBlocking { job.join() }
-      log.debug("Scheduled validation complete")
+      log.debug("Scheduled resource validation complete")
     } else {
-      log.debug("Scheduled validation disabled")
+      log.debug("Scheduled resource validation disabled")
     }
   }
 
+  @Scheduled(fixedDelayString = "\${keel.environment-check.frequency:PT1S}")
+  fun checkEnvironments() {
+    if (enabled) {
+      log.debug("Starting scheduled environment validation…")
+      publisher.publishEvent(ScheduledEnvironmentCheckStarting)
+
+      val job = launch {
+        deliveryConfigRepository
+          .launchForEachItem {
+            environmentPromotionChecker.checkEnvironments(it)
+          }
+      }
+
+      runBlocking { job.join() }
```

Comment:
This means we simultaneously check each resource then we wait for them all to finish, right?

Reply:
Yeah, because the `@Scheduled` annotation specifies a delay from when the function completes and when it will be called again, so here we want to wait until all the checks are complete. Otherwise subsequent checks would start to overlap each other.

Reply:
👍 just making sure I understood the coroutines `join()` bit